File: src/ipc.py
"""Unix Domain Socket IPC for waydeeper daemon management."""

import json
import logging
import os
import socket
import tempfile
import threading
from pathlib import Path
from typing import Callable, Any

logger = logging.getLogger(__name__)

RUNTIME_DIR = Path(tempfile.gettempdir()) / f"waydeeper-{os.getuid()}"


class DaemonSocket:
    """Socket server for daemon to accept commands."""

    def __init__(self, monitor: str, command_handler: Callable[[str, dict], Any]):
        self.monitor = monitor
        self.command_handler = command_handler
        self.socket_path = RUNTIME_DIR / f"{monitor}.sock"
        self.server = None
        self.running = False
        self.thread = None

    def start(self):
        """Start listening on socket."""
        RUNTIME_DIR.mkdir(parents=True, exist_ok=True)

        # Remove old socket if exists
        if self.socket_path.exists():
            try:
                # Test if another daemon is using it
                sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
                sock.settimeout(1)
                sock.connect(str(self.socket_path))
                sock.close()
                raise RuntimeError(
                    f"Another daemon already running for monitor {self.monitor}"
                )
            except (socket.error, ConnectionRefusedError):
                # Socket is stale, remove it
                self.socket_path.unlink()

        self.server = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
        self.server.bind(str(self.socket_path))
        self.server.listen(5)
        self.running = True

        self.thread = threading.Thread(target=self._accept_loop, daemon=True)
        self.thread.start()

        logger.info(f"IPC socket listening at {self.socket_path}")

    def stop(self):
        """Stop listening and cleanup socket."""
        self.running = False
        if self.server:
            self.server.close()
        if self.socket_path.exists():
            self.socket_path.unlink()
        logger.info("IPC socket stopped")

    def _accept_loop(self):
        """Accept and handle incoming connections."""
        while self.running:
            try:
                self.server.settimeout(1.0)
                conn, addr = self.server.accept()
                self._handle_connection(conn)
            except socket.timeout:
                continue
            except OSError:
                break

    def _handle_connection(self, conn: socket.socket):
        """Handle a single client connection."""
        try:
            conn.settimeout(5.0)
            data = b""
            while True:
                chunk = conn.recv(4096)
                if not chunk:
                    break
                data += chunk
                if b"\n" in data:
                    break

            if not data:
                return

            # Parse request
            request = json.loads(data.decode().strip())
            command = request.get("command", "")
            params = request.get("params", {})

            # Handle command
            result = self.command_handler(command, params)

            # Send response
            response = {"success": True, "result": result}
            conn.sendall(json.dumps(response).encode() + b"\n")

        except json.JSONDecodeError as e:
            response = {"success": False, "error": f"Invalid JSON: {e}"}
            conn.sendall(json.dumps(response).encode() + b"\n")
        except Exception as e:
            response = {"success": False, "error": str(e)}
            conn.sendall(json.dumps(response).encode() + b"\n")
        finally:
            conn.close()


class DaemonClient:
    """Client for communicating with daemon via socket."""

    def __init__(self, monitor: str):
        self.monitor = monitor
        self.socket_path = RUNTIME_DIR / f"{monitor}.sock"

    def is_running(self) -> bool:
        """Check if daemon is responsive."""
        if not self.socket_path.exists():
            return False
        try:
            response = self.send_command("PING", {}, timeout=2)
            return response.get("success", False)
        except:
            return False

    def send_command(
        self, command: str, params: dict = None, timeout: float = 5
    ) -> dict:
        """Send a command to the daemon and return response."""
        if not self.socket_path.exists():
            raise ConnectionRefusedError(
                f"Daemon not running for monitor {self.monitor}"
            )

        sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
        sock.settimeout(timeout)

        try:
            sock.connect(str(self.socket_path))

            request = {"command": command, "params": params or {}}
            sock.sendall(json.dumps(request).encode() + b"\n")

            # Receive response
            data = b""
            while True:
                chunk = sock.recv(4096)
                if not chunk:
                    break
                data += chunk
                if b"\n" in data:
                    break

            response = json.loads(data.decode().strip())
            return response

        finally:
            sock.close()

    def get_status(self) -> dict:
        """Get daemon status."""
        return self.send_command("STATUS")

    def stop_daemon(self) -> bool:
        """Request daemon to stop."""
        try:
            response = self.send_command("STOP")
            return response.get("success", False)
        except:
            return False


def list_running_daemons() -> list[str]:
    """List all monitors with responsive daemons."""
    if not RUNTIME_DIR.exists():
        return []

    running = []
    for sock_path in RUNTIME_DIR.glob("*.sock"):
        monitor = sock_path.stem
        client = DaemonClient(monitor)
        if client.is_running():
            running.append(monitor)
        else:
            # Cleanup stale socket
            try:
                sock_path.unlink()
            except:
                pass

    return running


def get_daemon_info(monitor: str) -> dict | None:
    """Get info about a running daemon."""
    client = DaemonClient(monitor)
    if not client.is_running():
        return None

    try:
        response = client.get_status()
        if response.get("success"):
            return response.get("result")
    except Exception as e:
        logger.warning(f"Failed to get info for {monitor}: {e}")

    return None


def stop_daemon(monitor: str, timeout: float = 5) -> bool:
    """Stop a daemon gracefully via socket."""
    client = DaemonClient(monitor)
    if not client.is_running():
        return False

    try:
        # Send stop command - if we get a success response, daemon will stop
        if client.stop_daemon():
            # Wait a moment for daemon to start shutting down
            import time

            time.sleep(0.5)
            # Verify daemon is no longer responsive
            for _ in range(int((timeout - 0.5) * 10)):
                if not client.is_running():
                    return True
                time.sleep(0.1)

        return False
    except Exception as e:
        logger.warning(f"Failed to stop daemon {monitor}: {e}")
        return False


def stop_all_daemons(timeout: float = 5) -> dict[str, bool]:
    """Stop all running daemons."""
    results = {}
    for monitor in list_running_daemons():
        results[monitor] = stop_daemon(monitor, timeout)
    return results

File: src/ipc.rs
use anyhow::{anyhow, Context, Result};
use serde::{Deserialize, Serialize};
use std::io::{Read, Write};
use std::os::unix::net::{UnixListener, UnixStream};
use std::path::PathBuf;
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;
use std::thread;
use std::time::Duration;

use crate::config;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct IpcRequest {
    pub command: String,
    #[serde(default)]
    pub params: serde_json::Value,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct IpcResponse {
    pub success: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub result: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

pub type CommandHandler =
    Arc<dyn Fn(&str, &serde_json::Value) -> Result<serde_json::Value> + Send + Sync>;

pub struct DaemonSocket {
    monitor: String,
    socket_path: PathBuf,
    running: Arc<AtomicBool>,
}

impl DaemonSocket {
    pub fn new(monitor: &str) -> Result<Self> {
        let runtime_dir = config::runtime_dir()?;
        let socket_path = runtime_dir.join(format!("{}.sock", monitor));

        Ok(Self {
            monitor: monitor.to_string(),
            socket_path,
            running: Arc::new(AtomicBool::new(false)),
        })
    }

    pub fn start(&mut self, handler: CommandHandler) -> Result<()> {
        let runtime_dir = config::runtime_dir()?;
        std::fs::create_dir_all(&runtime_dir)?;

        if self.socket_path.exists() {
            match UnixStream::connect(&self.socket_path) {
                Ok(_) => {
                    return Err(anyhow!(
                        "Another daemon already running for monitor {}",
                        self.monitor
                    ));
                }
                Err(_) => {
                    let _ = std::fs::remove_file(&self.socket_path);
                }
            }
        }

        let listener = UnixListener::bind(&self.socket_path)
            .with_context(|| format!("Failed to bind socket at {:?}", self.socket_path))?;

        listener
            .set_nonblocking(true)
            .context("Failed to set socket non-blocking")?;

        self.running.store(true, Ordering::SeqCst);
        let running = self.running.clone();

        thread::spawn(move || {
            while running.load(Ordering::SeqCst) {
                match listener.accept() {
                    Ok((mut stream, _)) => {
                        let handler = handler.clone();
                        thread::spawn(move || {
                            let _ = Self::handle_connection(&mut stream, &handler);
                        });
                    }
                    Err(ref error) if error.kind() == std::io::ErrorKind::WouldBlock => {
                        thread::sleep(Duration::from_millis(50));
                        continue;
                    }
                    Err(_) => break,
                }
            }
        });

        log::info!("IPC socket listening at {:?}", self.socket_path);
        Ok(())
    }

    fn handle_connection(
        stream: &mut UnixStream,
        handler: &CommandHandler,
    ) -> Result<()> {
        stream.set_read_timeout(Some(Duration::from_secs(5)))?;
        stream.set_write_timeout(Some(Duration::from_secs(5)))?;

        let mut data = Vec::new();
        let mut buffer = [0u8; 4096];

        loop {
            match stream.read(&mut buffer) {
                Ok(0) => break,
                Ok(bytes_read) => {
                    data.extend_from_slice(&buffer[..bytes_read]);
                    if data.contains(&b'\n') {
                        break;
                    }
                }
                Err(ref error) if error.kind() == std::io::ErrorKind::WouldBlock => {
                    if !data.is_empty() && data.contains(&b'\n') {
                        break;
                    }
                    thread::sleep(Duration::from_millis(10));
                    continue;
                }
                Err(error) => return Err(error.into()),
            }
        }

        if data.is_empty() {
            return Ok(());
        }

        let response = match serde_json::from_slice::<IpcRequest>(&data) {
            Ok(request) => match handler(&request.command, &request.params) {
                Ok(result) => IpcResponse {
                    success: true,
                    result: Some(result),
                    error: None,
                },
                Err(error) => IpcResponse {
                    success: false,
                    result: None,
                    error: Some(error.to_string()),
                },
            },
            Err(error) => IpcResponse {
                success: false,
                result: None,
                error: Some(format!("Invalid JSON: {}", error)),
            },
        };

        let response_data = serde_json::to_vec(&response)?;
        stream.write_all(&response_data)?;
        stream.write_all(b"\n")?;
        stream.flush()?;
        Ok(())
    }

    pub fn stop(&mut self) {
        self.running.store(false, Ordering::SeqCst);
        if self.socket_path.exists() {
            let _ = std::fs::remove_file(&self.socket_path);
        }
        log::info!("IPC socket stopped");
    }
}

impl Drop for DaemonSocket {
    fn drop(&mut self) {
        self.stop();
    }
}

pub struct DaemonClient {
    monitor: String,
    socket_path: PathBuf,
}

impl DaemonClient {
    pub fn new(monitor: &str) -> Result<Self> {
        let runtime_dir = config::runtime_dir()?;
        let socket_path = runtime_dir.join(format!("{}.sock", monitor));

        Ok(Self {
            monitor: monitor.to_string(),
            socket_path,
        })
    }

    pub fn is_running(&self) -> bool {
        if !self.socket_path.exists() {
            return false;
        }
        match self.send_command("PING", serde_json::Value::Null, Duration::from_secs(2)) {
            Ok(response) => response.success,
            Err(_) => false,
        }
    }

    pub fn send_command(
        &self,
        command: &str,
        params: serde_json::Value,
        timeout: Duration,
    ) -> Result<IpcResponse> {
        if !self.socket_path.exists() {
            return Err(anyhow!("Daemon not running for monitor {}", self.monitor));
        }

        let mut stream = UnixStream::connect(&self.socket_path)?;
        stream.set_read_timeout(Some(timeout))?;
        stream.set_write_timeout(Some(timeout))?;

        let request = IpcRequest {
            command: command.to_string(),
            params,
        };

        let request_data = serde_json::to_vec(&request)?;
        stream.write_all(&request_data)?;
        stream.write_all(b"\n")?;
        stream.flush()?;

        let mut data = Vec::new();
        let mut buffer = [0u8; 4096];

        loop {
            match stream.read(&mut buffer) {
                Ok(0) => break,
                Ok(bytes_read) => {
                    data.extend_from_slice(&buffer[..bytes_read]);
                    if data.contains(&b'\n') {
                        break;
                    }
                }
                Err(ref error) if error.kind() == std::io::ErrorKind::WouldBlock => break,
                Err(error) => return Err(error.into()),
            }
        }

        let response: IpcResponse = serde_json::from_slice(&data)?;
        Ok(response)
    }

    pub fn stop_daemon(&self) -> Result<bool> {
        let response =
            self.send_command("STOP", serde_json::Value::Null, Duration::from_secs(5))?;
        Ok(response.success)
    }
}

pub fn list_running_daemons() -> Result<Vec<String>> {
    let runtime_dir = match config::runtime_dir() {
        Ok(dir) => dir,
        Err(_) => return Ok(Vec::new()),
    };

    if !runtime_dir.exists() {
        return Ok(Vec::new());
    }

    let mut running = Vec::new();

    for entry in std::fs::read_dir(&runtime_dir)?.flatten() {
        let path = entry.path();
        if path.extension().is_some_and(|ext| ext == "sock") {
            if let Some(stem) = path.file_stem() {
                let monitor = stem.to_string_lossy().to_string();
                if let Ok(client) = DaemonClient::new(&monitor) {
                    if client.is_running() {
                        running.push(monitor);
                    } else {
                        let _ = std::fs::remove_file(&path);
                    }
                }
            }
        }
    }

    Ok(running)
}

pub fn stop_daemon(monitor: &str, timeout: Duration) -> Result<bool> {
    let client = DaemonClient::new(monitor)?;
    if !client.is_running() {
        return Ok(false);
    }

    if client.stop_daemon()? {
        thread::sleep(Duration::from_millis(500));
        let remaining = timeout.saturating_sub(Duration::from_millis(500));
        let steps = (remaining.as_millis() / 100) as u32;
        for _ in 0..steps {
            if !client.is_running() {
                return Ok(true);
            }
            thread::sleep(Duration::from_millis(100));
        }
    }

    Ok(false)
}

pub fn stop_all_daemons(
    timeout: Duration,
) -> Result<std::collections::HashMap<String, bool>> {
    let mut results = std::collections::HashMap::new();
    for monitor in list_running_daemons()? {
        results.insert(
            monitor.clone(),
            stop_daemon(&monitor, timeout).unwrap_or(false),
        );
    }
    Ok(results)
}
